node/src/types/block.rs

```diff
@@ -1111,6 +1113,226 @@ impl Item for BlockHeaderWithMetadata {
+    pub fn len(&self) -> u64 {
+        self.highest + 1 - self.lowest
+    }
+}
+
+impl Display for BlockHeadersBatchId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "block header batch {}..={}", self.highest, self.lowest)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct BlockHeadersBatch(Vec<BlockHeader>);
+
+impl BlockHeadersBatch {
+    /// Validates whether received batch is:
+    /// 1) highest block header from the batch has a hash is `latest_known.hash`
+    /// 2) a link of header[n].parent == header[n+1].hash is maintained
+    ///
+    /// Returns lowest block header from the batch or error if batch fails validation.
+    pub(crate) fn validate(
+        &self,
+        batch_id: &BlockHeadersBatchId,
+        earliest_known: &BlockHeader,
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Result<BlockHeader, BlockHeadersBatchValidationError> {
+        let highest_header = self
+            .0
+            .first()
+            .ok_or(BlockHeadersBatchValidationError::BatchEmpty)?;
+
+        if batch_id.len() != self.inner().len() as u64 {
+            return Err(BlockHeadersBatchValidationError::IncorrectLength {
+                expected: batch_id.len(),
+                got: self.inner().len() as u64,
+            });
+        }
+
+        // Check first header first b/c it's cheaper than verifying continuity.
+        let highest_hash = highest_header.hash(verifiable_chunked_hash_activation);
+        if &highest_hash != earliest_known.parent_hash() {
+            return Err(BlockHeadersBatchValidationError::HighestBlockHashMismatch {
+                expected: *earliest_known.parent_hash(),
+                got: highest_hash,
+            });
+        }
+
+        if !Self::is_continuous_and_descending(self.inner(), verifiable_chunked_hash_activation) {
+            return Err(BlockHeadersBatchValidationError::BatchNotContinuous);
+        }
+
+        self.0
+            .last()
+            .cloned()
+            .ok_or(BlockHeadersBatchValidationError::BatchEmpty)
+    }
+
+    /// Tries to create an instance of `BlockHeadersBatch` from a `Vec<BlockHeader>`.
+    ///
+    /// Returns `Some(Self)` if data passes validation, otherwise `None`.
+    pub fn from_vec(
+        batch: Vec<BlockHeader>,
+        requested_id: &BlockHeadersBatchId,
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Option<Self> {
+        match batch.first() {
+            Some(highest) => {
+                if highest.height() != requested_id.highest {
+                    error!(
+                        expected_highest=?requested_id.highest,
+                        got_highest=?highest,
+                        "unexpected highest block header"
+                    );
+                    return None;
+                }
+            }
+            None => {
+                error!("response cannot be an empty batch");
```

Comment:
Seems like the error logs in this method should be warnings only, since they could be invoked by a malicious peer sending invalid data.

Reply:
Done in [`36ded88` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/36ded88067816ec81e9ed1af5e997cf372fb127c)